File: zaiko/src/display/events.zig
// SDL2 event handling (T084-T090)
// Per maiko/src/sdl.c:process_SDLevents

const std = @import("std");
const errors = @import("../utils/errors.zig");
const sdl_backend = @import("sdl_backend.zig");
const sdl2 = @import("sdl2.zig");
const keyboard = @import("../io/keyboard.zig");
const mouse = @import("../io/mouse.zig");

/// Event type
pub const EventType = enum {
    KEY_PRESS,
    KEY_RELEASE,
    BUTTON_PRESS,
    BUTTON_RELEASE,
    MOTION,
    QUIT,
};

/// Event structure
pub const Event = union(EventType) {
    KEY_PRESS: keyboard.KeyboardEvent,
    KEY_RELEASE: keyboard.KeyboardEvent,
    BUTTON_PRESS: mouse.MouseEvent,
    BUTTON_RELEASE: mouse.MouseEvent,
    MOTION: mouse.MouseEvent,
    QUIT: void,
};

/// SDL keycode to Lisp keycode mapping
/// Per maiko/src/sdl.c:keymap[]
const KeyMapEntry = struct {
    lisp_keycode: u16,
    sdl_keycode: i32,
};

const KEYMAP: []const KeyMapEntry = &[_]KeyMapEntry{
    .{ .lisp_keycode = 0, .sdl_keycode = sdl2.SDLK_5 },
    .{ .lisp_keycode = 1, .sdl_keycode = sdl2.SDLK_4 },
    .{ .lisp_keycode = 2, .sdl_keycode = sdl2.SDLK_6 },
    .{ .lisp_keycode = 3, .sdl_keycode = sdl2.SDLK_e },
    .{ .lisp_keycode = 4, .sdl_keycode = sdl2.SDLK_7 },
    .{ .lisp_keycode = 5, .sdl_keycode = sdl2.SDLK_d },
    .{ .lisp_keycode = 6, .sdl_keycode = sdl2.SDLK_u },
    .{ .lisp_keycode = 7, .sdl_keycode = sdl2.SDLK_v },
    .{ .lisp_keycode = 8, .sdl_keycode = sdl2.SDLK_RIGHTPAREN },
    .{ .lisp_keycode = 8, .sdl_keycode = sdl2.SDLK_0 },
    .{ .lisp_keycode = 9, .sdl_keycode = sdl2.SDLK_k },
    .{ .lisp_keycode = 10, .sdl_keycode = sdl2.SDLK_MINUS },
    .{ .lisp_keycode = 11, .sdl_keycode = sdl2.SDLK_p },
    .{ .lisp_keycode = 12, .sdl_keycode = sdl2.SDLK_SLASH },
    .{ .lisp_keycode = 13, .sdl_keycode = sdl2.SDLK_KP_PERIOD },
    .{ .lisp_keycode = 14, .sdl_keycode = sdl2.SDLK_SCROLLLOCK },
    .{ .lisp_keycode = 15, .sdl_keycode = sdl2.SDLK_BACKSPACE },
    .{ .lisp_keycode = 16, .sdl_keycode = sdl2.SDLK_3 },
    .{ .lisp_keycode = 17, .sdl_keycode = sdl2.SDLK_2 },
    .{ .lisp_keycode = 18, .sdl_keycode = sdl2.SDLK_w },
    .{ .lisp_keycode = 19, .sdl_keycode = sdl2.SDLK_q },
    .{ .lisp_keycode = 20, .sdl_keycode = sdl2.SDLK_s },
    .{ .lisp_keycode = 21, .sdl_keycode = sdl2.SDLK_a },
    .{ .lisp_keycode = 22, .sdl_keycode = sdl2.SDLK_LEFTPAREN },
    .{ .lisp_keycode = 22, .sdl_keycode = sdl2.SDLK_9 },
    .{ .lisp_keycode = 23, .sdl_keycode = sdl2.SDLK_i },
    .{ .lisp_keycode = 24, .sdl_keycode = sdl2.SDLK_x },
    .{ .lisp_keycode = 25, .sdl_keycode = sdl2.SDLK_o },
    .{ .lisp_keycode = 26, .sdl_keycode = sdl2.SDLK_l },
    .{ .lisp_keycode = 27, .sdl_keycode = sdl2.SDLK_COMMA },
    .{ .lisp_keycode = 28, .sdl_keycode = sdl2.SDLK_QUOTE },
    .{ .lisp_keycode = 29, .sdl_keycode = sdl2.SDLK_RIGHTBRACKET },
    .{ .lisp_keycode = 31, .sdl_keycode = sdl2.SDLK_LALT },
    .{ .lisp_keycode = 32, .sdl_keycode = sdl2.SDLK_1 },
    .{ .lisp_keycode = 33, .sdl_keycode = sdl2.SDLK_ESCAPE },
    .{ .lisp_keycode = 34, .sdl_keycode = sdl2.SDLK_TAB },
    .{ .lisp_keycode = 35, .sdl_keycode = sdl2.SDLK_f },
    .{ .lisp_keycode = 36, .sdl_keycode = sdl2.SDLK_LCTRL },
    .{ .lisp_keycode = 37, .sdl_keycode = sdl2.SDLK_c },
    .{ .lisp_keycode = 38, .sdl_keycode = sdl2.SDLK_j },
    .{ .lisp_keycode = 39, .sdl_keycode = sdl2.SDLK_b },
    .{ .lisp_keycode = 40, .sdl_keycode = sdl2.SDLK_z },
    .{ .lisp_keycode = 41, .sdl_keycode = sdl2.SDLK_LSHIFT },
    .{ .lisp_keycode = 42, .sdl_keycode = sdl2.SDLK_PERIOD },
    .{ .lisp_keycode = 43, .sdl_keycode = sdl2.SDLK_SEMICOLON },
    .{ .lisp_keycode = 43, .sdl_keycode = sdl2.SDLK_COLON },
    .{ .lisp_keycode = 44, .sdl_keycode = sdl2.SDLK_RETURN },
    .{ .lisp_keycode = 45, .sdl_keycode = sdl2.SDLK_BACKQUOTE },
    .{ .lisp_keycode = 47, .sdl_keycode = sdl2.SDLK_RCTRL },
    .{ .lisp_keycode = 48, .sdl_keycode = sdl2.SDLK_r },
    .{ .lisp_keycode = 49, .sdl_keycode = sdl2.SDLK_t },
    .{ .lisp_keycode = 50, .sdl_keycode = sdl2.SDLK_g },
    .{ .lisp_keycode = 51, .sdl_keycode = sdl2.SDLK_y },
    .{ .lisp_keycode = 52, .sdl_keycode = sdl2.SDLK_h },
    .{ .lisp_keycode = 53, .sdl_keycode = sdl2.SDLK_8 },
    .{ .lisp_keycode = 54, .sdl_keycode = sdl2.SDLK_n },
    .{ .lisp_keycode = 55, .sdl_keycode = sdl2.SDLK_m },
    .{ .lisp_keycode = 56, .sdl_keycode = sdl2.SDLK_CAPSLOCK },
    .{ .lisp_keycode = 57, .sdl_keycode = sdl2.SDLK_SPACE },
    .{ .lisp_keycode = 58, .sdl_keycode = sdl2.SDLK_LEFTBRACKET },
    .{ .lisp_keycode = 59, .sdl_keycode = sdl2.SDLK_EQUALS },
    .{ .lisp_keycode = 60, .sdl_keycode = sdl2.SDLK_RSHIFT },
    .{ .lisp_keycode = 61, .sdl_keycode = sdl2.SDLK_F11 },
    .{ .lisp_keycode = 61, .sdl_keycode = sdl2.SDLK_PAUSE },
    .{ .lisp_keycode = 62, .sdl_keycode = sdl2.SDLK_HOME },
    .{ .lisp_keycode = 63, .sdl_keycode = sdl2.SDLK_PAGEUP },
    .{ .lisp_keycode = 64, .sdl_keycode = sdl2.SDLK_KP_EQUALS },
    .{ .lisp_keycode = 65, .sdl_keycode = sdl2.SDLK_KP_DIVIDE },
    .{ .lisp_keycode = 66, .sdl_keycode = sdl2.SDLK_F7 },
    .{ .lisp_keycode = 67, .sdl_keycode = sdl2.SDLK_F4 },
    .{ .lisp_keycode = 68, .sdl_keycode = sdl2.SDLK_F5 },
    .{ .lisp_keycode = 69, .sdl_keycode = sdl2.SDLK_KP_2 },
    .{ .lisp_keycode = 70, .sdl_keycode = sdl2.SDLK_KP_3 },
    .{ .lisp_keycode = 76, .sdl_keycode = sdl2.SDLK_KP_ENTER },
};

const KEYCODE_OFFSET: u16 = 0;

/// Map SDL keycode to Lisp keycode (T086)
/// Per maiko/src/sdl.c:map_key()
fn mapKey(sdl_keycode: i32) ?u16 {
    for (KEYMAP) |entry| {
        if (entry.sdl_keycode == sdl_keycode) {
            return entry.lisp_keycode;
        }
    }
    return null;
}

/// Extract modifier flags from SDL keymod
fn extractModifiers(sdl_mod: u16) u16 {
    var modifiers: u16 = 0;
    if (sdl_mod & sdl2.KMOD_SHIFT != 0) modifiers |= keyboard.Modifiers.SHIFT;
    if (sdl_mod & sdl2.KMOD_CTRL != 0) modifiers |= keyboard.Modifiers.CONTROL;
    if (sdl_mod & sdl2.KMOD_ALT != 0) modifiers |= keyboard.Modifiers.ALT;
    if (sdl_mod & sdl2.KMOD_GUI != 0) modifiers |= keyboard.Modifiers.META;
    return modifiers;
}

/// Poll SDL2 events (T084)
/// Per maiko/src/sdl.c:process_SDLevents()
pub fn pollEvents(
    display: *sdl_backend.DisplayInterface,
    key_queue: *keyboard.KeyEventQueue,
    mouse_state: *mouse.MouseState,
    allocator: std.mem.Allocator,
) errors.DisplayError!bool {
    _ = allocator;

    var sdl_event: sdl2.SDL_Event = undefined;

    // T084: Poll events in loop
    while (sdl2.SDL_PollEvent(&sdl_event) != 0) {
        switch (sdl_event.type) {
            sdl2.SDL_QUIT => {
                // Quit event - return true to signal quit
                return true;
            },

            // T085: Handle keyboard events
            sdl2.SDL_KEYDOWN => {
                const key_event = sdl_event.key;
                if (key_event.repeat != 0) {
                    // For key repeat, send UP before DOWN (C: "Lisp needs to see the UP transition before the DOWN transition")
                    if (mapKey(key_event.keysym.sym)) |lisp_keycode| {
                        const up_event = keyboard.KeyboardEvent{
                            .event_type = .KEY_RELEASE,
                            .keycode = lisp_keycode - KEYCODE_OFFSET,
                            .modifiers = extractModifiers(key_event.keysym.mod),
                            .timestamp = key_event.timestamp,
                        };
                        keyboard.enqueueKeyEvent(key_queue, up_event) catch {};
                    }
                }

                // T086: Translate keycode and T087: Deliver to queue
                if (mapKey(key_event.keysym.sym)) |lisp_keycode| {
                    const kb_event = keyboard.KeyboardEvent{
                        .event_type = .KEY_PRESS,
                        .keycode = lisp_keycode - KEYCODE_OFFSET,
                        .modifiers = extractModifiers(key_event.keysym.mod),
                        .timestamp = key_event.timestamp,
                    };
                    keyboard.enqueueKeyEvent(key_queue, kb_event) catch {};
                } else {
                    // No mapping found (C: printf("No mapping for key %s\n", SDL_GetKeyName(k)))
                    std.debug.print("No mapping for key {}\n", .{key_event.keysym.sym});
                }
            },

            sdl2.SDL_KEYUP => {
                const key_event = sdl_event.key;
                // T086: Translate keycode and T087: Deliver to queue
                if (mapKey(key_event.keysym.sym)) |lisp_keycode| {
                    const kb_event = keyboard.KeyboardEvent{
                        .event_type = .KEY_RELEASE,
                        .keycode = lisp_keycode - KEYCODE_OFFSET,
                        .modifiers = extractModifiers(key_event.keysym.mod),
                        .timestamp = key_event.timestamp,
                    };
                    keyboard.enqueueKeyEvent(key_queue, kb_event) catch {};
                }
            },

            // T088: Handle mouse events
            sdl2.SDL_MOUSEMOTION => {
                const motion_event = sdl_event.motion;
                // T089: Translate coordinates (divide by pixel_scale)
                const x = @divTrunc(@as(i32, @intCast(motion_event.x)), @as(i32, @intCast(display.pixel_scale)));
                const y = @divTrunc(@as(i32, @intCast(motion_event.y)), @as(i32, @intCast(display.pixel_scale)));

                // T090: Update mouse state
                mouse.updateMousePosition(mouse_state, x, y);
            },

            sdl2.SDL_MOUSEBUTTONDOWN => {
                const button_event = sdl_event.button;
                // T089: Translate coordinates
                const x = @divTrunc(@as(i32, @intCast(button_event.x)), @as(i32, @intCast(display.pixel_scale)));
                const y = @divTrunc(@as(i32, @intCast(button_event.y)), @as(i32, @intCast(display.pixel_scale)));

                // T090: Create mouse event
                // TODO: Use button_num for event delivery to Lisp
                _ = switch (button_event.button) {
                    sdl2.SDL_BUTTON_LEFT => @as(u8, 1),
                    sdl2.SDL_BUTTON_MIDDLE => @as(u8, 2),
                    sdl2.SDL_BUTTON_RIGHT => @as(u8, 3),
                    else => @as(u8, 0),
                };

                mouse.updateMousePosition(mouse_state, x, y);
                // Note: Button state is tracked in mouse_state, but event delivery to Lisp
                // would happen via a separate mechanism (similar to kb_trans)
            },

            sdl2.SDL_MOUSEBUTTONUP => {
                const button_event = sdl_event.button;
                // T089: Translate coordinates
                const x = @divTrunc(@as(i32, @intCast(button_event.x)), @as(i32, @intCast(display.pixel_scale)));
                const y = @divTrunc(@as(i32, @intCast(button_event.y)), @as(i32, @intCast(display.pixel_scale)));

                mouse.updateMousePosition(mouse_state, x, y);
                // Note: Button release handling similar to button down
            },

            else => {
                // Other events (window events, etc.) - ignore for now
            },
        }
    }
    return false;
}

/// Process mouse event (legacy interface)
/// Per contracts/display-interface.zig
pub fn processMouseEvent(display: *sdl_backend.DisplayInterface, sdl_event: *const anyopaque) mouse.MouseEvent {
    _ = display;
    const event: *const sdl2.SDL_Event = @ptrCast(@alignCast(sdl_event));

    switch (event.type) {
        sdl2.SDL_MOUSEBUTTONDOWN => {
            return mouse.MouseEvent{
                .event_type = .BUTTON_PRESS,
                .button = @intCast(event.button.button),
                .x = event.button.x,
                .y = event.button.y,
                .modifiers = 0, // TODO: Get modifier state
                .timestamp = event.button.timestamp,
            };
        },
        sdl2.SDL_MOUSEBUTTONUP => {
            return mouse.MouseEvent{
                .event_type = .BUTTON_RELEASE,
                .button = @intCast(event.button.button),
                .x = event.button.x,
                .y = event.button.y,
                .modifiers = 0, // TODO: Get modifier state
                .timestamp = event.button.timestamp,
            };
        },
        sdl2.SDL_MOUSEMOTION => {
            return mouse.MouseEvent{
                .event_type = .MOTION,
                .button = 0,
                .x = event.motion.x,
                .y = event.motion.y,
                .modifiers = 0, // TODO: Get modifier state
                .timestamp = event.motion.timestamp,
            };
        },
        else => {
            // Unknown mouse event type, return motion with zeros
            return mouse.MouseEvent{
                .event_type = .MOTION,
                .button = 0,
                .x = 0,
                .y = 0,
                .modifiers = 0,
                .timestamp = 0,
            };
        },
    }
}

File: taiko/src/display/events.ts
// Browser event handling
// Translates browser keyboard/mouse events to Lisp events

/**
 * Keycode mapping from browser keycodes to Lisp keycodes
 * Based on maiko/src/sdl.c keymap array (74 entries)
 * Maps browser keycodes -> SDL keycodes -> Lisp keycodes
 *
 * The keymap is structured as pairs: [lisp_keycode, sdl_keycode]
 * We need to reverse this to: browser_keycode -> lisp_keycode
 */
const BROWSER_TO_SDL_KEYMAP: Map<number, number> = new Map([
    // Basic keys
    [8, 42],      // Backspace -> SDLK_BACKSPACE
    [9, 34],      // Tab -> SDLK_TAB
    [13, 44],     // Enter -> SDLK_RETURN
    [27, 33],     // Escape -> SDLK_ESCAPE
    [32, 57],     // Space -> SDLK_SPACE
    [33, 63],     // Page Up -> SDLK_PAGEUP
    [34, 70],     // Page Down -> SDLK_KP_3 (PGDN)
    [35, 90],     // End -> SDLK_END
    [36, 62],     // Home -> SDLK_HOME
    [37, 0],      // Left Arrow -> (mapped separately)
    [38, 0],      // Up Arrow -> (mapped separately)
    [39, 0],      // Right Arrow -> (mapped separately)
    [40, 0],      // Down Arrow -> (mapped separately)
    [45, 89],     // Insert -> SDLK_INSERT
    [46, 0],      // Delete -> (mapped separately)

    // Function keys
    [112, 0],     // F1
    [113, 0],     // F2
    [114, 0],     // F3
    [115, 67],    // F4 -> SDLK_F4
    [116, 68],    // F5 -> SDLK_F5
    [117, 66],    // F6 -> SDLK_F7 (UTIL2)
    [118, 0],     // F7
    [119, 0],     // F8
    [120, 80],    // F9 -> SDLK_F9
    [121, 61],    // F10 -> SDLK_F11
    [122, 61],    // F11 -> SDLK_F11
    [123, 91],    // F12 -> SDLK_F12

    // Number row
    [48, 8],      // 0 -> SDLK_0
    [49, 32],     // 1 -> SDLK_1
    [50, 17],     // 2 -> SDLK_2
    [51, 16],     // 3 -> SDLK_3
    [52, 1],      // 4 -> SDLK_4
    [53, 0],      // 5 -> SDLK_5
    [54, 2],      // 6 -> SDLK_6
    [55, 4],      // 7 -> SDLK_7
    [56, 53],     // 8 -> SDLK_8
    [57, 22],     // 9 -> SDLK_9

    // Letters (A-Z) - mapped to lowercase SDL codes
    [65, 21],     // A -> SDLK_a
    [66, 39],     // B -> SDLK_b
    [67, 37],     // C -> SDLK_c
    [68, 5],      // D -> SDLK_d
    [69, 3],      // E -> SDLK_e
    [70, 35],     // F -> SDLK_f
    [71, 50],     // G -> SDLK_g
    [72, 52],     // H -> SDLK_h
    [73, 23],     // I -> SDLK_i
    [74, 38],     // J -> SDLK_j
    [75, 9],      // K -> SDLK_k
    [76, 26],     // L -> SDLK_l
    [77, 55],     // M -> SDLK_m
    [78, 54],     // N -> SDLK_n
    [79, 25],     // O -> SDLK_o
    [80, 11],     // P -> SDLK_p
    [81, 19],     // Q -> SDLK_q
    [82, 48],     // R -> SDLK_r
    [83, 20],     // S -> SDLK_s
    [84, 49],     // T -> SDLK_t
    [85, 6],      // U -> SDLK_u
    [86, 7],      // V -> SDLK_v
    [87, 18],     // W -> SDLK_w
    [88, 24],     // X -> SDLK_x
    [89, 51],     // Y -> SDLK_y
    [90, 40],     // Z -> SDLK_z

    // Punctuation and symbols
    [186, 43],    // Semicolon -> SDLK_SEMICOLON
    [187, 59],    // Equals -> SDLK_EQUALS
    [188, 27],    // Comma -> SDLK_COMMA
    [189, 10],    // Minus -> SDLK_MINUS
    [190, 42],    // Period -> SDLK_PERIOD
    [191, 12],    // Slash -> SDLK_SLASH
    [192, 45],    // Backtick -> SDLK_BACKQUOTE
    [219, 58],    // Left Bracket -> SDLK_LEFTBRACKET
    [220, 0],     // Backslash -> (mapped separately)
    [221, 29],    // Right Bracket -> SDLK_RIGHTBRACKET
    [222, 28],    // Quote -> SDLK_QUOTE

    // Keypad
    [96, 98],     // KP 0 -> SDLK_KP_0
    [97, 81],     // KP 1 -> SDLK_KP_7
    [98, 82],     // KP 2 -> SDLK_KP_8
    [99, 83],     // KP 3 -> SDLK_KP_9
    [100, 84],    // KP 4 -> SDLK_KP_4
    [101, 85],    // KP 5 -> SDLK_KP_5
    [102, 87],    // KP 6 -> SDLK_KP_6
    [103, 81],    // KP 7 -> SDLK_KP_7
    [104, 82],    // KP 8 -> SDLK_KP_8
    [105, 83],    // KP 9 -> SDLK_KP_9
    [106, 95],    // KP Multiply -> SDLK_KP_MULTIPLY
    [107, 96],    // KP Plus -> SDLK_KP_PLUS (approximate)
    [109, 96],    // KP Minus -> SDLK_KP_MINUS
    [110, 13],    // KP Period -> SDLK_KP_PERIOD
    [111, 65],    // KP Divide -> SDLK_KP_DIVIDE
]);

/**
 * SDL to Lisp keycode mapping
 * Based on maiko/src/sdl.c keymap array structure
 * Maps SDL keycode -> Lisp keycode
 */
const SDL_TO_LISP_KEYMAP: Map<number, number> = new Map([
    // Create reverse mapping from SDL keymap pairs
    // [lisp_keycode, sdl_keycode] -> sdl_keycode -> lisp_keycode
    [42, 42],     // SDLK_PERIOD -> 42
    [1, 1],       // SDLK_4 -> 1
    [2, 2],       // SDLK_6 -> 2
    [3, 3],       // SDLK_e -> 3
    [4, 4],       // SDLK_7 -> 4
    [5, 5],       // SDLK_d -> 5
    [6, 6],       // SDLK_u -> 6
    [7, 7],       // SDLK_v -> 7
    [8, 8],       // SDLK_0 -> 8
    [9, 9],       // SDLK_k -> 9
    [10, 10],     // SDLK_MINUS -> 10
    [11, 11],     // SDLK_p -> 11
    [12, 12],     // SDLK_SLASH -> 12
    [13, 13],     // SDLK_KP_PERIOD -> 13
    [14, 14],     // SDLK_SCROLLLOCK -> 14
    [15, 15],     // SDLK_BACKSPACE -> 15
    [16, 16],     // SDLK_3 -> 16
    [17, 17],     // SDLK_2 -> 17
    [18, 18],     // SDLK_w -> 18
    [19, 19],     // SDLK_q -> 19
    [20, 20],     // SDLK_s -> 20
    [21, 21],     // SDLK_a -> 21
    [22, 22],     // SDLK_9 -> 22
    [23, 23],     // SDLK_i -> 23
    [24, 24],     // SDLK_x -> 24
    [25, 25],     // SDLK_o -> 25
    [26, 26],     // SDLK_l -> 26
    [27, 27],     // SDLK_COMMA -> 27
    [28, 28],     // SDLK_QUOTE -> 28
    [29, 29],     // SDLK_RIGHTBRACKET -> 29
    [31, 31],     // SDLK_LALT -> 31
    [32, 32],     // SDLK_1 -> 32
    [33, 33],     // SDLK_ESCAPE -> 33
    [34, 34],     // SDLK_TAB -> 34
    [35, 35],     // SDLK_f -> 35
    [36, 36],     // SDLK_LCTRL -> 36
    [37, 37],     // SDLK_c -> 37
    [38, 38],     // SDLK_j -> 38
    [39, 39],     // SDLK_b -> 39
    [40, 40],     // SDLK_z -> 40
    [41, 41],     // SDLK_LSHIFT -> 41
    [43, 43],     // SDLK_SEMICOLON -> 43
    [44, 44],     // SDLK_RETURN -> 44
    [45, 45],     // SDLK_BACKQUOTE -> 45
    [47, 47],     // SDLK_RCTRL -> 47
    [48, 48],     // SDLK_r -> 48
    [49, 49],     // SDLK_t -> 49
    [50, 50],     // SDLK_g -> 50
    [51, 51],     // SDLK_y -> 51
    [52, 52],     // SDLK_h -> 52
    [53, 53],     // SDLK_8 -> 53
    [54, 54],     // SDLK_n -> 54
    [55, 55],     // SDLK_m -> 55
    [56, 56],     // SDLK_CAPSLOCK -> 56
    [57, 57],     // SDLK_SPACE -> 57
    [58, 58],     // SDLK_LEFTBRACKET -> 58
    [59, 59],     // SDLK_EQUALS -> 59
    [60, 60],     // SDLK_RSHIFT -> 60
    [61, 61],     // SDLK_F11 -> 61
    [62, 62],     // SDLK_HOME -> 62
    [63, 63],     // SDLK_PAGEUP -> 63
    [64, 64],     // SDLK_KP_EQUALS -> 64
    [65, 65],     // SDLK_KP_DIVIDE -> 65
    [66, 66],     // SDLK_F7 -> 66
    [67, 67],     // SDLK_F4 -> 67
    [68, 68],     // SDLK_F5 -> 68
    [69, 69],     // SDLK_KP_2 -> 69
    [70, 70],     // SDLK_KP_3 -> 70
    [76, 76],     // SDLK_KP_ENTER -> 76
    [80, 80],     // SDLK_F9 -> 80
    [81, 81],     // SDLK_KP_7 -> 81
    [82, 82],     // SDLK_KP_8 -> 82
    [83, 83],     // SDLK_KP_9 -> 83
    [84, 84],     // SDLK_KP_4 -> 84
    [85, 85],     // SDLK_KP_5 -> 85
    [86, 86],     // SDLK_LALT -> 86
    [87, 87],     // SDLK_KP_6 -> 87
    [89, 89],     // SDLK_INSERT -> 89
    [90, 90],     // SDLK_END -> 90
    [91, 91],     // SDLK_F12 -> 91
    [92, 92],     // SDLK_PRINTSCREEN -> 92
    [93, 93],     // SDLK_MODE -> 93
    [94, 94],     // SDLK_KP_1 -> 94
    [95, 95],     // SDLK_KP_MULTIPLY -> 95
    [96, 96],     // SDLK_KP_MINUS -> 96
    [97, 97],     // SDLK_HELP -> 97
    [98, 98],     // SDLK_KP_0 -> 98
]);

/**
 * Translate browser keycode to Lisp keycode
 *
 * @param browserKeycode Browser keycode (from KeyboardEvent.keyCode)
 * @returns Lisp keycode, or browser keycode if no mapping found
 */
export function translateKeycode(browserKeycode: number): number {
    // First map browser keycode to SDL keycode
    const sdlKeycode = BROWSER_TO_SDL_KEYMAP.get(browserKeycode);
    if (sdlKeycode === undefined || sdlKeycode === 0) {
        // No mapping found, return browser keycode as fallback
        return browserKeycode;
    }

    // Then map SDL keycode to Lisp keycode
    const lispKeycode = SDL_TO_LISP_KEYMAP.get(sdlKeycode);
    if (lispKeycode === undefined) {
        // No mapping found, return SDL keycode as fallback
        return sdlKeycode;
    }

    return lispKeycode;
}

/**
 * Handle keyboard event
 *
 * @param event Browser keyboard event
 * @returns Lisp keycode, or null if event should be ignored
 */
export function handleKeyboardEvent(event: KeyboardEvent): number | null {
    const lispKeycode = translateKeycode(event.keyCode);
    return lispKeycode;
}

/**
 * Handle mouse event
 * Translates window coordinates to display coordinates
 *
 * @param event Browser mouse event
 * @param canvas Canvas element
 * @param pixelScale Pixel scale factor (default: 1)
 * @returns Mouse coordinates and button, or null if invalid
 */
export function handleMouseEvent(
    event: MouseEvent,
    canvas: HTMLCanvasElement,
    pixelScale: number = 1
): { x: number; y: number; button: number } | null {
    const rect = canvas.getBoundingClientRect();
    const x = Math.floor((event.clientX - rect.left) / pixelScale);
    const y = Math.floor((event.clientY - rect.top) / pixelScale);

    return {
        x,
        y,
        button: event.button,
    };
}
